File: minesweeper_bfs.py

#Time complexity : O(MN)
#Space complexity : O(MN) 
from collections import deque
class Solution:
    def __init__(self,dirs=None):
        self.dirs = [(0,-1),(0,1),(-1,0),(1,0),(-1,-1),(1,-1),(-1,1),(1,1)]
        self.m = None
        self.n = None
    
    def count_mines(self, board,point):
        
        cp = 0
        for item in self.dirs:
            r = point[0]+item[0]
            c = point[1]+item[1]
                    
            if (r>=0 and r < self.m) and (c>=0 and c < self.n) and board[r][c]=='M': #count only mines
                cp+=1
        
       
        return cp
            
        
    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        
        if board[click[0]][click[1]] == 'M':
            board[click[0]][click[1]] = 'X'
            return board
            
        self.m = len(board)
        self.n = len(board[0])
        queue = deque()
        board[click[0]][click[1]] = 'B'
        queue.append((click[0],click[1]))
        
        while(queue):
            popped = queue.popleft()
            mines = self.count_mines(board,popped)
            if mines>0:
                board[popped[0]][popped[1]] = str(mines)
            else:
                for direct in self.dirs:
                    r = popped[0]+direct[0]
                    c = popped[1]+direct[1]
                    
                    
                    if (r>=0 and r < self.m) and (c>=0 and c < self.n) and board[r][c]=='E': #only empty cells to be added
                        board[r][c]='B'
                        queue.append((r,c))
            
        return board
                    
        

File: minesweeperBfs.java
class Solution {
    int[][] dirs; int m; int n;
    public char[][] updateBoard(char[][] board, int[] click) {
        if(board == null || board.length==0) return board;
        dirs = new int[][]{{0,1}, {0, -1}, {-1, 0}, {1,0}, {-1,-1}, {1,1}, {-1, 1}, {1, -1}};
        m = board.length; n = board[0].length;
        if(board[click[0]][click[1]]=='M'){
            board[click[0]][click[1]] = 'X';
            return board;
        }
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{click[0], click[1]});
        board[click[0]][click[1]] = 'B';
        while(!q.isEmpty()) {
            int[] curr = q.poll();
            int countmines = countMines(board, curr[0], curr[1]);
            if(countmines>0) {
                board[curr[0]][curr[1]] = (char)(countmines +'0');
            } else {
                for(int[] dir: dirs){
                    int r = dir[0] + curr[0];
                    int c = dir[1] + curr[1];
                    if(r>=0 && r<m && c>=0 && c< n && board[r][c] == 'E'){
                        q.add(new int[]{r,c});
                        board[r][c] = 'B';
                    }
                }
            }
        }
        return board;
    }
    
    private int countMines(char[][]board, int i, int j) {
        int count = 0;
        for(int [] dir: dirs) {
            int r = dir[0] + i;
            int c = dir[1] + j;
            if(r>= 0 && r<m && c>=0 && c<n && board[r][c]=='M') count++;
        }
        
        return count;
    }
}